// zscale/funcs/scale.js
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var Utils = require('./../utils');
var d3 = require('d3-scale');
var getDomain = Utils.getDomain,
    getRange = Utils.getRange;


var funcs = {
	log: d3.scaleLog,
	pow: d3.scalePow,
	linear: d3.scaleLinear
};

var checkType = function checkType(d, valueType) {
	if (!d) return false;
	if (valueType === 'scale') return true;
	if ((typeof d === 'undefined' ? 'undefined' : _typeof(d)) !== 'object') return false;
	var type = d.type || valueType;
	var value = d.type ? d.value : d;
	if (type !== 'scale') return false;
	if (!value || !value.range || !value.domain) return false;
	return true;
};

var getFunc = function getFunc(o) {
	var type = o.easing || o.scale || 'linear';
	if (!type) return console.log(o, 'easing参数不存在');
	var func = funcs[type];
	if (!func) return console.log(o, '函数不存在');
	return func;
};

var generate = function generate(o) {
	var d3Func = getFunc(o);
	var domain = getDomain(o);
	var range = getRange(o);
	return d3Func().domain(domain).range(range);
};

module.exports = { generate: generate, checkType: checkType };
